pkg/kvstore/consul.go

```diff
@@ -304,59 +314,67 @@ func (c *ConsulClient) setMaxL3n4AddrID(maxID uint32) error {
 }
 
 func (c *ConsulClient) GASNewL3n4AddrID(basePath string, baseID uint32, lAddrID *types.L3n4AddrID) error {
-
-	setIDtoL3n4Addr := func(lockPair *consulAPI.KVPair) error {
-		defer c.KV().Release(lockPair, nil)
-		lAddrID.ID = types.ServiceID(baseID)
+	setIDtoL3n4Addr := func(id uint32) error {
+		lAddrID.ID = types.ServiceID(id)
 		keyPath := path.Join(basePath, strconv.FormatUint(uint64(lAddrID.ID), 10))
 		if err := c.SetValue(keyPath, lAddrID); err != nil {
 			return err
 		}
-		return c.setMaxL3n4AddrID(baseID + 1)
+		return c.setMaxL3n4AddrID(id + 1)
 	}
 
 	session, _, err := c.Session().CreateNoChecks(nil, nil)
 	if err != nil {
 		return err
 	}
 
-	beginning := baseID
-	for {
-		log.Debugf("Trying to acquire a new free ID %d", baseID)
-		keyPath := path.Join(basePath, strconv.FormatUint(uint64(baseID), 10))
+	acquireFreeID := func(firstID uint32, incID *uint32) (bool, error) {
+		log.Debugf("Trying to acquire a new free service ID %d", *incID)
+		keyPath := path.Join(basePath, strconv.FormatUint(uint64(*incID), 10))
 
 		lockPair := &consulAPI.KVPair{Key: GetLockPath(keyPath), Session: session}
 		acq, _, err := c.KV().Acquire(lockPair, nil)
 		if err != nil {
-			return err
+			return false, err
 		}
+		defer c.KV().Release(lockPair, nil)
 
 		if acq {
 			svcKey, _, err := c.KV().Get(keyPath, nil)
 			if err != nil {
-				c.KV().Release(lockPair, nil)
-				return err
+				return false, err
 			}
 			if svcKey == nil {
-				return setIDtoL3n4Addr(lockPair)
+				return false, setIDtoL3n4Addr(*incID)
 			}
 			var consulL3n4AddrID types.L3n4AddrID
 			if err := json.Unmarshal(svcKey.Value, &consulL3n4AddrID); err != nil {
-				c.KV().Release(lockPair, nil)
-				return err
+				return false, err
 			}
 			if consulL3n4AddrID.ID == 0 {
 				log.Infof("Recycling Service ID %d", baseID)
-				return setIDtoL3n4Addr(lockPair)
+				return false, setIDtoL3n4Addr(*incID)
 			}
-			c.KV().Release(lockPair, nil)
 		}
-		baseID++
-		if baseID > common.MaxSetOfServiceID {
-			baseID = common.FirstFreeServiceID
+
+		*incID++
+		if *incID > common.MaxSetOfServiceID {
```

Comment:
Similar to my comments above - would be helpful to explain what the logic of trying to acquire a service ID is succinctly - i.e., keep incrementing, loop around, and retry until we've hit the first ID, which means that there are no IDs left available.